### PR TITLE
Implement datetime64 support for Time

### LINKED ISSUE #6428: Support creating astropy.time.Time from numpy.datetime64 and pandas datetime columns
It would be great if this just worked:

```
import numpy as np
from astropy.time import Time

t =  np.array(['2017-01-01 20:00', '2017-01-01 21:00'], dtype='datetime64[ms]')
t = Time(t)
```

But right now, this results in a value error:
```
ValueError: No time format was given, and the input is not unique
```

### PR DESCRIPTION
**Pull request by vn-ki**
I just wanted to know if this is the right way to go.

This PR implements `numpy.datetime64` support for `Time`. This method first converts `datetime64` to string and then convert to `Time`.

The other methods (using unix time) gave errors at nanosecond precision.

Closes #6428.

### CODE REVIEW

**Comment by pllim** [comment]
@vn-ki , if this is to solve an existing issue, please cross-reference it. Thank you!

**Comment by vn-ki** [comment]
@pllim I have updated to include the issue.

**Comment by vn-ki** [comment]
@taldcroft @mhvk If anyone of you could take a look it would be great.

I just want to know whether you are okay with the approach taken with this PR.

Converting from datetime64 to Time through string seems a bit weird, but I didn't find a better way. Also with this, `precision` is set based on the precision of `datetime64` object (Incase of array as input, as of now, it puts the `precision` depending on the first element of the array). 

Also, if `Time`  has only 1 element, the `value` of `datetime64` is a float. It's because `.item()` is called in the below snippet. If you have any suggestion to circumvent this, that would be great too.

https://github.com/astropy/astropy/blob/5f9294d8135c6d1cdd4e8fc2e744f208d30b8f47/astropy/time/core.py#L621-L625

File: astropy/time/formats.py [hunk 1]
```diff
@@ -942,6 +942,31 @@ class TimeYearDayTime(TimeISO):
                 '{year:d}:{yday:03d}'))
 
 
+class TimeDatetime64(TimeISOT):
+    name = 'datetime64'
+    def _check_val_type(self, val1, val2):
```
Comment (taldcroft) [inline astropy/time/formats.py]: This should just check the `dtype` of the `val1` itself, not every element.  For `datetime64` this is a built-in numpy dtype so every element is sure to have the same dtype.  This is different from `datetime`, where `val1` will be an array of objects (dtype=`'O'`).

File: astropy/time/formats.py [hunk 2]
```diff
@@ -942,6 +942,31 @@ class TimeYearDayTime(TimeISO):
...
+        if not all(isinstance(val, np.datetime64) for val in val1.flat):
+            raise TypeError('Input values for {0} class must be '
+                            'datetime64 objects'.format(self.name))
+        return val1, None
+
+    def set_jds(self, val1, val2):
+        if val1.dtype.name in ['datetime64[M]', 'datetime64[Y]']:
+            val1 = val1.astype('M8[D]')
+
+        val1 = val1.astype('<U30')
```
Comment (taldcroft) [inline astropy/time/formats.py]: You should be able to use `val1.astype('S')`.  This saves space and will just have as many characters as are needed.

File: astropy/time/formats.py [hunk 3]
```diff
@@ -942,6 +942,31 @@ class TimeYearDayTime(TimeISO):
...
+
+        if '.' in val1.item(0):
```
Comment (taldcroft) [inline astropy/time/formats.py]: `precision` should not be set here, this is an input to the object initializer and user can expect either the documented default or what they set when creating the object.

File: astropy/time/formats.py [hunk 4]
```diff
@@ -942,6 +942,31 @@ class TimeYearDayTime(TimeISO):
...
+            self.precision = len(val1.item(0).split('.')[-1])
+
+        super(TimeDatetime64, self).set_jds(val1, val2)
```
Comment (taldcroft) [inline astropy/time/formats.py]: Use `super().set_jds(val1, val2)` for Python 3.  Yay!

File: astropy/time/formats.py [hunk 5]
```diff
@@ -942,6 +942,31 @@ class TimeYearDayTime(TimeISO):
...
+    def _check_val_type(self, val1, val2):
+        # Note: don't care about val2 for this class`
+        if not all(isinstance(val, np.datetime64) for val in val1.flat):
+            raise TypeError('Input values for {0} class must be '
+                            'datetime64 objects'.format(self.name))
+        return val1, None
+
+    def set_jds(self, val1, val2):
+        if val1.dtype.name in ['datetime64[M]', 'datetime64[Y]']:
+            val1 = val1.astype('M8[D]')
```
Comment (taldcroft) [inline astropy/time/formats.py]: Use `datetime64` not `M8`.  It just makes it easier to read.

File: astropy/time/formats.py [hunk 6]
```diff
@@ -969,6 +969,32 @@ class TimeYearDayTime(TimeISO):
...
+
+        val1 = val1.astype('S')
+
+        super().set_jds(val1, val2)
+
+    @property
+    def value(self):
+        precision = self.precision
+        self.precision = 9
+        ret = super(TimeDatetime64, self).value
```
Comment (taldcroft) [inline astropy/time/formats.py]: Use `super().value` as per previous comment.

File: CHANGES.rst [hunk 7]
```diff
@@ -170,6 +170,7 @@ astropy.time
   These methods are similar to those in the Python standard library
   `time` package and provide flexible input and output formatting. [#7323]
 
+- Added `numpy.datetime64` format class for ``Time`` class. [#7361]
```
Comment (taldcroft) [inline CHANGES.rst]: Say
```
- Added ``datetime64`` format to the ``Time`` class to support working with 
  `numpy.datetime64` dtype arrays. [#7361]
```

**Comment by taldcroft** [comment]
@vn-ki - do you know what's happening with the failing tests?

**Comment by vn-ki** [comment]
@taldcroft I missed a comma. It should pass now.

File: CHANGES.rst (hunk 7 above)
Comment (vn-ki) [reply CHANGES.rst]: Docs build failing.

```
../CHANGES.rst:173: WARNING: py:obj reference target not found: numpy.datetime64
```

Comment (vn-ki) [reply CHANGES.rst]: I'll do 
```
``numpy.datetime64``
```

Comment (mhvk) [reply CHANGES.rst]: Needs a twiddle: `~numpy.datetime64`.

Comment (mhvk) [reply CHANGES.rst]: Sorry, in `CHANGES.rst`, the double back ticks are the way to go, so keep your change!

**Comment by astrofrog** [comment]
@mhvk @taldcroft - can we merge this in time for v3.1 now that things are passing?

**Pull request merged**